horovod/torch/__init__.py

```diff
@@ -129,15 +131,48 @@ def broadcast_parameters(params, root_rank):
     `model.named_parameters()`, or `model.parameters()`.
 
     Arguments:
-        params: The list of parameters to broadcast.
+        params: One of the following:
+            - list of parameters to broadcast
+            - dict of parameters to broadcast
+            - torch.optim.Optimizer whose state will be broadcast
         root_rank: The rank of the process from which parameters will be
                    broadcasted to all other processes.
     """
     if isinstance(params, dict):
         params = sorted(params.items())
-    else:
+    elif isinstance(params, list):
         # support both named_parameters() and regular parameters()
         params = [p if isinstance(p, tuple) else (None, p) for p in params]
+    elif isinstance(params, torch.optim.Optimizer):
+        # Newly created optimizers will not have their state initialized, so
+        # do that initialization here
+        state_dict = params.state_dict()
+        if len(state_dict['state']) == 0:
+            for group in params.param_groups:
+                for p in group['params']:
+                    p.grad = torch.autograd.Variable(
+                        p.data.new(p.size()).zero_())
+            params.step()
+            state_dict = params.state_dict()
+
+        new_params = []
+        occurrences = collections.defaultdict(int)
+
+        # Groups are unordered, but their params will be distinct
+        for group in state_dict['param_groups']:
+            # The params list here is ordered by the layers in the model
+            for pid in group['params']:
+                param_state = state_dict['state'][pid]
+                for name, p in param_state.items():
+                    # Some parameter names may appear more than once, in which
+                    # case we ensure they have a unique identifier defined by
+                    # their order
+                    occurrences[name] += 1
+                    name = '%s.%d' % (str(name), occurrences[name])
```

Comment:
Is there guarantee here that if there are two parameters named `xyz` in each process (on different groups/params), they will always be traversed in the same order?